### PR TITLE
ARROW-4776: [C++] Add DictionaryBuilder constructor which takes a dictionary array

### PR DESCRIPTION
**Pull request by bkietz**
DictionaryBuilder can now be constructed with a user provided dictionary.

This is also available through MakeBuilder

### CODE REVIEW

**Comment by bkietz** [comment]
this PR should also resolve https://issues.apache.org/jira/browse/ARROW-4671

File: cpp/src/arrow/array/builder_dict.cc [hunk 1]
```diff
@@ -52,9 +52,8 @@ struct UnifyDictionaryValues {
   Status Visit(const DataType&, void* = nullptr) {
     // Default implementation for non-dictionary-supported datatypes
     std::stringstream ss;
```
Comment (xhochy) [inline cpp/src/arrow/array/builder_dict.cc]: You're no longer using `ss`.
```suggestion
```

**Review by xhochy — COMMENTED** [review]
Minor nitpick otherwise LGTM

**Comment by bkietz** [comment]
@kszucs [postgres hash mismatch in the JDK job](https://travis-ci.org/apache/arrow/jobs/505453250#L225), could you restart it?

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**